### DISCUSSION TITLE
Incompatible variable override involving inner classes and multiple inheritance

### DISCUSSION THREAD

**Discussion by Pyprohly** [microsoft/pyright#3182] [Q&A] [answered]
To better organise the many attributes that a class may have, I sometimes like to group related attributes together in an inner-class object. Occasionally I would need to add fields to the inner object in a subclass, so I would involve the inner-class in inheritance too.

```python
# pyright: strict

from __future__ import annotations

class BaseComment:
    submission: Submission

    class Submission:
        ...

class BaseExtraSubmissionFieldsComment(BaseComment):
    class Submission(BaseComment.Submission):
        # Add more attributes, etc.
        ...


class Comment(BaseComment):
    ...

class ExtraSubmissionFieldsComment(Comment, BaseExtraSubmissionFieldsComment):
    ...
```

In this situation I have, a subclass only adds new fields, all of which are related and should be grouped in the inner-class.

(The reason I have the ‘base’ versions of the classes is because I don’t want to duplicate the attribute fields—and the rather messy `def __init__` code—in both sync and async versions of the classes. I realise that the field `submission` would need to be typed as `BaseComment.Submission` in derived classes but I don’t have a better design right now and I don’t want to commit to having `submission` be an immutable field at the moment.)

Since the changes in version 1.1.229 (specifically #2936), the forth class definition errors with:

```text
error: Base classes for class "ExtraSubmissionFieldsComment" define variable "Submission" in incompatible way (reportIncompatibleVariableOverride)
```

I’m not really sure if this error message is unsound or if my code needs fixing somehow?

**Comment by erictraut** [comment, marked as answer]
I think the error message is correct here. You can choose to ignore it if you know that you have defined your inner classes appropriately such that they do not create problems for multiple inheritance. This practice looks very fragile to me, so I wouldn't recommend this approach.

**Reply by Pyprohly** [reply]
I know. It’s probably for the best that I abandon this cute attribute grouping idea and simply give each attribute a common prefix instead.
